kube-client/src/api/util.rs

```diff
@@ -34,6 +42,49 @@ impl Api<Node> {
     }
 }
 
+type CSRStatusType = &'static str;
+static APPROVED: CSRStatusType = "Approved";
+static DENIED: CSRStatusType = "Denied";
+
+impl Api<CertificateSigningRequest> {
+    /// Approve the specified CertificateSigningRequest.
+    pub async fn approve(&self, name: &str, pp: &PatchParams) -> Result<CertificateSigningRequest> {
+        self.patch_approval(name, pp, &APPROVED).await
+    }
+
+    /// Deny the specified CertificateSigningRequest.
+    pub async fn deny(&self, name: &str, pp: &PatchParams) -> Result<CertificateSigningRequest> {
+        self.patch_approval(name, pp, &DENIED).await
+    }
+
+    /// Partially update approval of the specified CertificateSigningRequest.
+    async fn patch_approval(
+        &self,
+        name: &str,
+        pp: &PatchParams,
+        approval_type: &CSRStatusType,
+    ) -> Result<CertificateSigningRequest> {
+        let patch = serde_json::json!({"status": CertificateSigningRequestStatus {
+            certificate:None,
+            conditions: Some(vec![CertificateSigningRequestCondition {
+                type_: approval_type.to_string(),
+                last_update_time: None,
+                last_transition_time: None,
+                message: Some(format!("{} {}", approval_type, "by kube-rs client")),
+                reason: Some("kube-rsClient".to_string()),
+                status: "True".to_string()
+            }])
+        }});
+
+        let mut req = self
+            .request
+            .patch_subresource("approval", name, pp, &Patch::Strategic(&patch))
```

Comment:
we are also taking `PatchParams` in this fn, but always using `Patch::strategic`, so this is not ideal.

Reply:
I guess we should use Patch::Merge then. Please let me know if I am wrong!

Reply:
The problem is that we let users set their own `PatchParams`, but unlike `Api::patch`, they don't actually supply a `Patch` themselves, so it's unlike anything we have and doesn't match conventions.

Options:

1. name the function `set_approval`, and just not expose the `PatchParams`
2. take a `Patch<T>` in `patch_approval` instead of a specific struct (but have a doc example use a `CertificateSigningRequest`

1 is arguably more ergonomic, but it makes choices that we probably should not be making in `kube-client`.
Otoh, 2 is just a dumb wrapper for `Api::patch_subresource("approval", ...)`, but this plays into goals for generic subresource handling nicer down the line.

Personally, I think we should follow the dumb approach here because it matches how we do `get_scale` / `patch_scale` / `get_status` / `patch_status`. If we do the dumb options, then we can also do:

- `Api::get_approval`
- `Api::replace_approval`

AFAIKT we could lock down the types for these 2 (like `Scale`) because it looks they return a full `CertificateSigningRequest` (so the subresource is kind of the same struct as itself, the part it cares about is just a subset of it..).

---

side-notes on generic subresource handling: background info https://github.com/kube-rs/k8s-pb/issues/4 and discussion in https://github.com/Arnavion/k8s-openapi/pull/98#issuecomment-853439773 (although this has been put on hold for a bit to try to tackle it direct in k8s-pb)

If the Approval subresource is not its own struct but reusing the root one, then the subresource behaves like status and not scale. This makes one of the original subresource plans kind of awkward: `Api<CertificateSigningRequest>::subresource<CertificateSigningRequest>::patch()` is rather ugly, so it's perhaps better to stick with the longer names like `Api<CertificateSigningRequest>::patch_approval()` long term - and find a way to generate this in the future)

Reply:
Got it, thank you for the explanation!

Reply:
Thanks a lot - and for your patience during the new year :-)
Everything in the PR is great and happy to merge it as is.

Just have one question left. Do you think there's a benefit to adding `Api::get_approval` while we are here?
AFAIKT it would just do the same thing as `Api::get` for the CSR (because it's the same struct), but technically [kubernetes supports the GET verb on the approval subresource](https://github.com/kube-rs/k8s-pb/blob/d60271637aac5bb854d529f99d5e6a23154acd06/k8s-pb-codegen/openapi/transformed.json#L804-L809).

Reply:
Thank you for reviewing during the holidays! I'd be happy to add `Api<CSR>::get_approval` if that's useful, but isn't it just another wrapper of `get`:
```rust
pub async fn get_approval(
        &self,
        name: &str) -> Result<CertificateSigningRequest> {
      self.get(name)
}
```

Reply:
I think if we do it, it should be a wrapper for your new `get_subresource("approval", name)` because then it's up to the apiserver whether or not it's actually handled differently (rather than us making an alias for consistency).

But I don't think we need to do it if this actually returns the same thing as `get(name)` which i suspect it does. If you are able to verify, then that would be good to know (and maybe good to leave a comment about for the next person).

Reply:
Interestingly, while both `get(csr_name)` and `get_subresource("approval", csr_name)` returns a CertificateSigningRequest, the returned CSR differs in two places: 
1. `resource_version`
2. `status.certificate`: it's `none` for `get(name)` but not none for `get_subresource("approval", csr_name)`.

Reply:
Ah, interesting! Then we should definitely also export an alias `get_approval` as an alias for `get_subresource("approval", ...)`

Reply:
Just added that!

Reply:
Looks great. Left a minor follow-up comment on the test (which is still using `.get`).